recipes/libxslt/all/conanfile.py

```diff
@@ -33,167 +38,161 @@ class LibxsltConan(ConanFile):
         "plugins": False,
     }
 
-    @property
-    def _source_subfolder(self):
-        return "source_subfolder"
-
     @property
     def _settings_build(self):
         return getattr(self, "settings_build", self.settings)
 
-    @property
-    def _option_names(self):
-        return [name for name in self.default_options.keys() if name not in ["shared", "fPIC"]]
-
     def export_sources(self):
-        for patch in self.conan_data.get("patches", {}).get(self.version, []):
-            self.copy(patch["patch_file"])
+        export_conandata_patches(self)
 
     def config_options(self):
         if self.settings.os == "Windows":
             del self.options.fPIC
 
     def configure(self):
         if self.options.shared:
-            del self.options.fPIC
-        del self.settings.compiler.libcxx
-        del self.settings.compiler.cppstd
+            self.options.rm_safe("fPIC")
+        self.settings.rm_safe("compiler.cppstd")
+        self.settings.rm_safe("compiler.libcxx")
+
+    def layout(self):
+        basic_layout(self, src_folder="src")
 
     def requirements(self):
-        self.requires("libxml2/2.9.14")
+        self.requires("libxml2/2.10.3", transitive_headers=True, transitive_libs=True)
```

Comment:
``transitive_libs=True`` is an exceptional situation, it shouldn't be necessary in the majority of cases. I think its usage deserves a short comment why it is there.

Reply:
It's the opposite actually, or I've missed something in the meaning of these traits. As soon as headers of a dependency libB are public in libA (ie headers of libB are exposed in public headers of libA), libA is very likely using symbols which are not inlined in headers of libB but rather symbols for which implementation lies in binary of libA.

Reply:
If a consumer is using public headers of library A, and _only_ public headers and symbols of library A, in a way that causes the consumer's compiled translation units, to directly reference symbols from library B - without the consumer's knowledge, then `transitive_libs` would need to be `True`, specifically to cover the case where both libraries are shared.

However, there are also a lot of situations where the transitive header exposure causes the consumer to _have_ to be aware and directly and explicitly use the transitive dependency - thus, the consumer should be directly linking against t he library, and `transitive_libs` should not be true.

I think the `test_package` in this instance is a good example: the test package is _directly_ invoking code from libxml2 (because it's including its headers) - so it falls on the consumer (at the very least) to link against libxml2. 

having inspected the headers of libxlt2, I can see that we are in the second case, and not the first - thus `transitive_libs` should not be true, and the test_package CMakeLists.txt of the consumer should be linking against libxml2 (because it is the consumer that is using it too - this is in line with good include/link what you use practice).

Reply:
Take a look at `libxslt/xsltutils.h` public header. It defines a macro `IS_XSLT_NAME` which calls `xmlStrEqual`. `xmlStrEqual` implementation is in libxml2 binaries, so libs must be public.

Reply:
> It's the opposite actually, or I've missed something in the meaning of these traits. As soon as headers of a dependency libB are public in libA (ie headers of libB are exposed in public headers of libA), libA is very likely using symbols which are not inlined in headers of libB but rather symbols for which implementation lies in binary of libB, so transitive_libs must be True. And by the way transitive_headers=True, transitive_libs=True is the proper translation of CMake target_link_libraries(libA PUBLIC libB).

It is not something always ``True`` or always ``False.`` 
The cases that needs ``transitive_libs=True`` should use it. As @jcar87 explained, it happens sometimes, when symbols are used in headers. But it doesn't apply always. And when it doesn't apply it is very important to not define it "just in case", because it can be harmful and produce incorrect linkage in the consumers, which is exactly one of the things that traits try to solve.
So ``transitive_headers=True`` => ``transitive_libs=True`` is False in the general case, and shouldn't be added by default.
For the cases that is necessary, sure, it can be added, the trait exist for this reason. It would be great to have 1-2 lines of comment explaining the case (the header that is causing the symbol usage, or some small explanation).

Reply:
I've not said it was always True or always False, but that most of the time when headers are public, lib is also public. This is how CMake works -through target abstraction (a target is public, interface or private)-, this is how pkgconfig works (it doesn't even have the concept of a dependency which would have public headers and private lib AFAIK).

In an ideal world, each contributor would dig into all the details of public headers of a lib to check whether dependencies headers are included (transitive_headers=True if they are included) and whether pre-compiled symbols are used (transivite_libs=True).
Obviously it's really hard to check that, and actually even upstream CMake config files and pkgconfig files don't even model such granularity (in CMake when a target is public, both headers and libs are propagated).
What I want to say is that you should not ask to contributors to remove transitive_libs=True when there is transitive_headers=True. Instead you should ask to explain why transitive_libs is not True when transitive_headers is True, because this specific case is unusual, not the other one.

Reply:
Also I would like to mention that conan team has added transitive_headers=True in several recipes (without transitive_libs=True), and realized afterwards that it was not sufficient and transitive_libs was also required. It's a waste of time and c3i resources. Just add transitive_libs=True by default when headers are public, and optimize afterwards eventually.
As I said it's tedious to be really sure that libs are not transitive when headers are included unless you have a very good knowledge of the library.

Reply:
Thanks for looking into it and most of all adding a comment for prosperity :) 

I understand this was is more work but it will allow us to test our hypothesis from the client's perspective and maybe this will end up being in change in 3.0 🤷 but we need to measure it 

Since as you pointed out we dont know (and it's next to impossible to know without doing IMO)